### PR TITLE
Initial release of open source module

### PR DESCRIPTION
**Pull request by ryanckoch**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,192 @@
...
+Be sure you have the correct Terraform version (0.10.x), you can choose the binary here:
+- https://releases.hashicorp.com/terraform/
+
+## Usage
+There are multiple examples included in the [examples](./examples/) folder but simple usage is as follows:
+
+```hcl
+module "gke" {
+  source                     = "github.com/terraform-google-modules/terraform-google-kubernetes-engine"
+  credentials_path           = "${local.credentials_file_path}"
```
Comment (morgante) [inline README.md]: Let's remove credentials_path everywhere!

File: README.md [hunk 2]
```diff
@@ -1,2 +1,191 @@
...
+module "gke" {
+  source                     = "github.com/terraform-google-modules/terraform-google-kubernetes-engine"
+  project_id                 = "<PROJECT ID>"
+  cluster_name               = "gke-test-1"
+  region                     = "us-central1"
+  network                    = "vpc-01"
+  subnetwork                 = "us-central1-01"
+  ip_range_pods              = "us-central1-01-gke-01-pods"
+  ip_range_services          = "us-central1-01-gke-01-services"
+  node_service_account       = "project-service-account@<PROJECT ID>.iam.gserviceaccount.com"
```
Comment (morgante) [inline README.md]: Service account should be configurable only on a node_pool level so I can have two node pools with different service accounts. Use the service account of the first node pool for the empty "default" node pool.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,191 @@
...
+#### Outputs
+Please refer the /outputs.tf file for the outputs that you can get with the `terraform output` command
+
+## Infrastructure
+The resources/services/activations/deletions that this module will create/trigger are:
+- Create a GKE cluster with the provided addons
+- Create GKE Node Pool(s) with provided configuration and attach to cluster
+- Replace the default kube-dns configmap if `stub_domains` are provided
+- Activate network policy if `network_policy` is true
+- Add `ip-masq-agent` configmap with provided `masq_non_masquerade_cidrs` if `network_policy` is true or `masq_config_enabled` is true
```
Comment (morgante) [inline README.md]: Change `masq_non_masquerade_cidrs` to `non_masquerade_cidrs`.

File: examples/deploy_service/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,59 @@
+/**
```
Comment (morgante) [inline examples/deploy_service/outputs.tf]: Don't need all outputs for all examples, just actually useful things (ex. endpoint for service).

File: examples/deploy_service/main.tf [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    session_affinity = "ClientIP"
+
+    port {
+      port        = 8080
+      target_port = 80
+    }
+
+    type = "ClusterIP"
```
Comment (morgante) [inline examples/deploy_service/main.tf]: Let's deploy as a load balancer, easier to see results.

File: variables.tf [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+variable "horizontal_pod_autoscaling" {
+  description = "Enable horizontal pod autoscaling addon"
+  default     = true
+}
+
+variable "http_load_balancing" {
+  description = "Enable httpload balancer addon"
+  default     = false
```
Comment (morgante) [inline variables.tf]: Should default to enabled.

File: variables.tf [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+variable "node_version" {
+  description = "The Kubernetes version of the node pools. Defaults kubernetes version (master) variable. Must set the same as master at initial creation."
+  default     = ""
+}
+
+variable "horizontal_pod_autoscaling" {
+  description = "Enable horizontal pod autoscaling addon"
+  default     = true
```
Comment (morgante) [inline variables.tf]: Should default to disabled.

File: variables.tf [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+variable "http_load_balancing" {
+  description = "Enable httpload balancer addon"
+  default     = false
+}
+
+variable "kubernetes_dashboard" {
+  description = "Enable kubernetes dashboard addon"
+  default     = true
```
Comment (morgante) [inline variables.tf]: Should be disabled by default.

File: variables.tf [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+variable "ip_masq_config_enabled" {
+  description = "Enable IP Masquerade Agent Configmap. Automatically enables if network_policy enabled."
+  default     = false
+}
+
+variable "ip_masq_non_masquerade_cidrs" {
+  type        = "list"
+  description = "List of strings in CIDR notation that specify the IP address ranges that do not use IP masquerading."
+  default     = ["10.0.0.0/8"]
```
[the remark this answers is not in the document]
Comment (morgante) [reply variables.tf]: @endzyme Good catch. @ryanckoch We should update the default to include all of RFC1918 space, as documented here: https://cloud.google.com/kubernetes-engine/docs/how-to/ip-masquerade-agent#how_ipmasq_works

[the remark this answers is not in the document]
Comment (endzyme) [reply variables.tf]: I would recommend against appending new rules to RFC1918. We had just hit a use case for supplying subsets of networks. Just for clarity to the user on declarative nature of the config.

[the remark this answers is not in the document]
Comment (morgante) [reply variables.tf]: Sounds good, let's make sure to include a note in the docs about this.

**Comment by morgante** [comment]
@ryanckoch I've moved this code to a branch on the shared repo so we can collaborate, could you open a new PR from it: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine

File: README.md (hunk 3 above)
Comment (ryanckoch) [reply README.md]: Done.

File: examples/deploy_service/outputs.tf (hunk 4 above)
Comment (ryanckoch) [reply examples/deploy_service/outputs.tf]: Done.

**Pull request closed without merging**